![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

# Proyecto: Innovación Tecnológica en Bodoque Bank

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### Cuerpo Docente:

- Profesor: Pablo Badilla, Ignacio Meza De La Jara
- Auxiliar: Sebastián Tinoco
- Ayudante: Diego Cortez M., Felipe Arias T.

*Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir.*


----

## Reglas

- Fecha de entrega: 01/06/2021
- **Grupos de 2 personas.**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Estrictamente prohibida la copia. 
- Pueden usar cualquier material del curso que estimen conveniente.

# Motivación

![](https://www.geekmi.news/__export/1621969098810/sites/debate/img/2021/05/25/juan-carlos-bodoque-1.jpg_554688468.jpg)

Juan Carlos Bodoque, el famoso periodista y empresario, decidió diversificar su portafolio de negocios y crear su propio banco. Después de varios años de investigar y analizar el mercado financiero, finalmente logró fundar su entidad bancaria con el objetivo de ofrecer a sus clientes una experiencia personalizada y confiable en sus transacciones financieras. 

Sin embargo, con las nuevas tecnologías, aparecen nuevos desafíos para la joven entidad bancaria. Por ello, Bodoque decide invertir en un equipo de expertos en tecnología y finanzas, para que Bodoque Bank implemente las últimas innovaciones en seguridad y servicio al cliente para garantizar la satisfacción y fidelización de sus clientes.

El primer objetivo de la entidad bancaria será la detección de potenciales clientes fraudulentos, para ello Bodoque Bank les hace entrega de un extenso dataset en el que se registran las actividades que han realizado sus clientes durante los últimos meses. Uno de los puntos que resaltan al pasar el conjunto de datos es que el nombre de los usuarios está protegido y que consideren cada una de las filas como una muestra independiente de la otra.

Tras la solicitud, uno de los mayores accionistas del banco llamado Mario Hugo, les sugiere que al momento de realizar el proyecto tomen las siguientes consideraciones:

- Realicen un análisis exhaustivo de cada variable, ya que los directores tienden a ser quisquillosos con puntos o definiciones sin algún fundamento claro.
- Deben ser muy claros de cómo encontrar los outliers y si es posible expliquen cómo están determinando que un cliente es fraudulento.
- Es muy probable que existan clientes con diferentes comportamientos bancarios, por lo que sería muy buena decisión de separar a los clientes en grupos etarios.


### Definición Formal del Problema

El dataset con el que se trabajará en este primer proyecto será [_The Bank Account Fraud (BAF)_](https://www.kaggle.com/datasets/sgpjesus/bank-account-fraud-dataset-neurips-2022). Este consiste en un dataset para evaluar métodos de detección de fraudes bancarios, el cuál según sus autores es:

- Realista: Se basa en un actualizado dataset de casos reales de detección de fraude bancario.
- Sesgado: Poseen distintos tipos de bias.
- Desbalanceado: Clase positiva extremadamente pequeña.
- Dinámico: Presenta datos temporales y cambios de distribución.
- Preserva la privacidad: Protege a los potenciales clientes a través de la aplicación de privacidad diferencial.

El proyecto tiene por objetivo evaluar los conocimientos adquiridos en la primera mitad del curso, consistente en manejo de datos tabulares (I/O, manipulación, agregaciones, merge y visualizaciones) más la primera parte de modelos consistente en preprocesamiento de datos y detección de anomalías. Por ende, el proyecto consiste en dos tareas principales:

1. Generar un Análisis exploratorio de Datos (EDA) que describa completamente el dataset.
2. Buscar anomalías de forma automatizada.

**Importante:** Esta permitido el uso de librerías externas a las vistas en clases para profundizar aún mas en los análisis. Sin embargo, al momento de utilizar cualquier método deberán explicar qué hace y el porqué de su aplicación.


---

## Secciones requeridas en la entrega

La siguiente lista detalla las secciones que debe contener su notebook para resolver el proyecto. Es importante que al momento de desarrollar cada una de las secciones, estas sean escritas en un formato tipo **informe**, donde describan detalladamente cada uno de los puntos realizados.

### 1. Introducción [0.5 Puntos]

Elaborar una breve introducción con todo lo necesario para entender qué realizarán durante su proyecto. La idea es que describan de manera formal el proyecto con sus propias palabras y logren describir algunos aspectos básicos tanto del dataset como del análisis a realizar sobre los datos.

Por lo anterior, en esta sección ustedes deberán ser capaces de:

- Describir la tarea asociada al dataset.
- Describir brevemente los datos de entrada que les provee el problema.
- Plantear hipótesis de cómo podrían abordar el problema.

### 2. Lectura y Manejo de Datos [1 Punto]

Este punto tiene por objetivo evaluar la capacidad de leer y manejar distintas fuentes de datos y unirlas.
Para esto, se les provee de 3 datasets distintos: 

- `df_1.parquet` que contiene una cierta cantidad de ejemplos.
- `df_2.parquet` que contiene el resto de los ejemplos.
- `df_email_phone.parquet` que contiene características exclusivas relacionadas a los emails y teléfonos de los ejemplos.

Tanto `df_1` como `df_2` son dataframes que contienen conjuntos (no necesariamente disjuntos) de ejemplos obtenidos a partir de distintas fuentes (la primera más confiable que la segunda)  mientras que `df_email_phone` contiene nuevas características que deben ser fusionadas con el conjunto de ejemplos.
Dicho esto, para este punto se le solicita:

1. Cargar los datasets usando pandas.
2. Explorar superficialmente los datsets por medio de `.head(5)` e `.info()` y reportar número de filas y columnas de cada dataset.
3. Unir los datasets con los conjuntos de ejemplos en un único dataset.
4. Combinar el datasets con ejemplos con las nuevas características usando la estrategia `outer`.
5. Explorar nulos y duplicados de datos (i.e., generar conteos de cada uno y ver algunos de sus ejemplos), explicar su origen y entregar una solución sencilla a estos casos.
6. Limpiar datos y reportar modificaciones a los datasets originales (cuántos ejemplos se eliminaron, cuántos ejemplos quedaron finalmente, etc...).
7. Realizar una segmentación etaria de los datos creando una columna llamada `segmentacion_etaria` para clasificar a los diferentes usuarios. Para la segmentación utilice los siguientes rangos:
   - Joven (Menor a 18 años)
   - Adulto-Joven (18 - 26 años)
   - Adulto (27- 59 años)
   - Persona Mayor (60 años o mas).


### 3. Análisis Exploratorio de Datos [1 Punto]

Esta sección se busca evaluar la capacidad de generar reportes y conclusiones a partir de los patrones observados en estos.

- Utilizando `pandas-profiling` (nombre de la librería: `ydata-profiling`), explore:

    * Analizar cantidad de datos nulos, tipos de datos, duplicados, distribuciones de las variables a través de histogramas.
    * Generar visualizaciones de las interacciones (como por ejemplo, una scatter matrix) en las distintas variables.
    * Ver las correlaciones entre las distintas variables y los valores faltantes de cada una de estas. 
    * Reportar los patrones interesantes observados. Ejemplos:
        - Se observó que las variables numéricas A, B y C siguen una distribución normal / exponencial / logarítmica / no presenta distribución conocida debido a que (usar información sobre el descripción de cada variable)....
        - La mayoría de los ejemplos de las variables I, J y K tienen valor 0.
        - La variable M presenta gran cantidad de outliers.
        - Se observó que las variables categóricas E y F están bien balanceadas / presentan un gran desbalance.
        - Del gráfico de correlaciones se puede inferir que X, Y y Z tienen una correlación alta/baja entre ellas.



### 4. Preparación de Datos [1 Punto]

*Esta sección consiste en generar los distintos pasos para preparar sus datos con el fin de luego poder crear su modelo.*

Dentro de los aspectos minimos que se les solicitarán en este desarrollo estan:

- Definir preprocesadores para datos categóricos y ordinales.
- Setear las transformaciones en un `ColumnTransformer`.
- Transformar todo el datset.

**Notas**: 

- Las variables `id` y `fraud_bool` no deben ser transformadas, pero si pasadas a la siguiente etapa. (Estudiar la transformación `passthrough`).
- La salida de la transformación debe ser un pandas dataframe. Para esto, investiguen acerca de la nueva [`set_output` API](https://scikit-learn.org/stable/auto_examples/miscellaneous/plot_set_output.html). Si se levanta un error por matriz sparse, desactiven su uso en la transformación correspondiente.

### 5. Visualización en baja dimensionalidad [1 Punto]

Programar un script para proyectar los datos en baja dimensionalidad usando `PCA` o `UMAP`.

**Notas**: 

- Para realizar las proyecciones no consideren la variable `fraud_bool`, ya que esto provocaría un _data leakage_ (i.e., usar información del futuro para analizar el mismo fenómeno que quieren descubrir).
- Utilice un muestreo significativo del dataframe (10k a 25k ejemplos) para generar las proyecciones. De lo contrario, probablemente el proyector no sea capaz de asignar todos los recursos necesarios y lance alguna excepción.

Luego, por cada segmento etario generado en la sección 2, implemente un gráfico de dispersión (idealmente en plotly express) la proyección en 2D generada y coloree cada punto según la etiqueta `fraud_bool`.

Reportar si existen patrones y/o relaciones interesantes con respecto a variables de interés.


### 6. Explorar Anomalías [1.5 Puntos]

Proponer una técnica para detectar clientes fraudulentos en base a detección de anomalías. 
Al igual que el punto anterior, la técnica que proponga debe ser aplicada por separado en los diferentes segmentos etarios descritos en el punto 2 (es decir, debe entrenar un detector de anomalías para cada uno de estos grupos). 

Luego, al igual que el punto anterior, genere un gráfico de dispersión con las proyecciones 2D en donde el coloreado sea ahora las etiquetas predichas por el detector de anomalías.

Por último, usando queries de numpy/pandas, calcule dos ratios (P y R):

- P: Cantidad de ejemplos predichos correctamente como fraude / cantidad total de datos predichos como fraude.
- R: Cantidad de ejemplos predichos correctamente como fraude / cantidad total de ejemplos que eran realmente fraude.


De los resultados obtenidos responda las siguientes preguntas:

- ¿Qué significan los ratios y sus valores?
- ¿Qué tan correctas fueron las predicciones realizadas por su modelo según los ratios calculados?
- ¿Son coherentes los resultados obtenidos?.
- ¿Cómo se comportan los casos de fraudes para los diferentes rangos etarios?

Justifique cada una de sus respuestas.

---

## Esquema de la Tarea


Pueden usar el siguiente esquema para organizar la tarea (y borrar todo lo anterior).
Obviamente **no deben limitarse a lo que está escrito en esta**: puede incrementar en caso de más técnicas y obviar algunas partes en caso que alguna y otro punto no aplique a su problema.

Pueden borrar las instrucciones anteriores y quedarse solo con lo que viene a continuación.

---
<br>

<br>

<br>



# Proyecto

### Equipo:

- Tomás Ceballos
- Pablo Angel


### Link de repositorio de GitHub: `\<http://....\>`




## 1. Introducción


**Describir la tarea asociada**

**Describir los datos de entrada que se recibieron**

**Plantear hipótesis de cómo abordar el problema**

## 2. Lectura y Manejo de Datos




### 2.1 Cargar datos en dataframes

In [1]:
# Librerias de manejo de datos
import numpy as np
import pandas as pd 

from IPython.display import display

#Libreria para visualizar
#!pip install --upgrade plotly
import plotly.figure_factory as ff
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go

#Librerias de sklearn
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.pipeline import Pipeline
from sklearn.decomposition import PCA

In [2]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = '/content/drive/My Drive/Proyecto_01'
except: 
    path = '/Users/tomasceballos/Universidad/Lab/Proyecto1'
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


In [3]:
# Leemos el primer archivo parquet
df_1 = pd.read_parquet(path + '/df_1.parquet')

# Leemos el segundo archivo parquet
df_2 = pd.read_parquet(path + '/df_2.parquet')

# Leemos el tercer archivo parquet
df_email_phone = pd.read_parquet(path + '/df_email_phone.parquet')

### 2.2 Exploración inicial

Ahora para poder analizar de forma automatizada todos los dataframes, vamos a crear una función que realice un EDA de manera general. Con todos los datos que nos entrega esta función no va a ser necesario utilizar el método *info()* de los dataframes. Con esta función se van a entregar los siguientes datos:

* Número de filas y columnas del dataset
* Los nombres de las columnas
* Las primeras 5, últimas 5, y 5 filas aleatorias del dataset
* La descripción numérica del dataset, es decir, una descripción general de las columnas numéricas de este
* Tipo de datos por columna
* Valores nulos por columna
* Valores nulos en total
* Valores únicos por columna



In [4]:
def exploratory_data_analysis(dataframe):
    pd.options.display.max_rows = 100
    
    print('1.- El DataFrame tiene', dataframe.shape[0],'filas y', dataframe.shape[1], 'columnas\n')
    print('2.- El DataFrame esta compuesto por las siguientes columnas: ', list(dataframe.columns),'\n')
    print('3.- Ejemplos de filas del DataFrame:\n')
    print('Primeras 5 filas:')
    display(dataframe.head(5))
    print('\n Últimas 5 filas:')
    display(dataframe.tail(5))
    print('\n Muestreo aleatorio de 5 filas:')
    display(dataframe.sample(5))
    print('\n 4.- Descripción numérica del Dataframe:\n')
    display(dataframe.describe())
    print('\n 5.- Tipo de datos por columna:')
    display(dataframe.dtypes)
    print('\n 6.- Cantidad de valores nulos por columna:')
    display(dataframe.isna().sum())
    print('\n 7.- Cantidad de valores nulos en total:')
    display(dataframe.isna().sum().sum())
    print('\n 8.- Cantidad de valores únicos por columna:')
    display(dataframe.nunique())

In [5]:
exploratory_data_analysis(df_1)

1.- El DataFrame tiene 389782 filas y 29 columnas

2.- El DataFrame esta compuesto por las siguientes columnas:  ['id', 'fraud_bool', 'income', 'prev_address_months_count', 'current_address_months_count', 'customer_age', 'days_since_request', 'intended_balcon_amount', 'payment_type', 'zip_count_4w', 'velocity_6h', 'velocity_24h', 'velocity_4w', 'bank_branch_count_8w', 'employment_status', 'credit_risk_score', 'housing_status', 'bank_months_count', 'has_other_cards', 'proposed_credit_limit', 'foreign_request', 'source', 'session_length_in_minutes', 'device_os', 'keep_alive_session', 'device_fraud_count', 'month', 'x1', 'x2'] 

3.- Ejemplos de filas del DataFrame:

Primeras 5 filas:


id  fraud_bool  income  prev_address_months_count  \
104446  729517           0     0.7                         -1   
269483  149585           0     0.8                         -1   
4102     64486           0     0.9                         -1   
351767  825283           0     0.5                         -1   
126377    8308           0     0.9                         -1   

        current_address_months_count  customer_age  days_since_request  \
104446                           305            60            0.030059   
269483                           140            50            0.015659   
4102                             171            50            0.001409   
351767                            85            30            0.027292   
126377                            39            30            0.010945   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
104446               -1.599455           AC           990  ...   
269483                3.951994           AA          1269  ...   
4102                 28.159779           AB          4430  ...   
351767               -1.310498           AB          1698  ...   
126377               -1.450972           AC           569  ...   

        proposed_credit_limit  foreign_request    source  \
104446                  500.0                0  INTERNET   
269483                  200.0                0  INTERNET   
4102                   1500.0                0  INTERNET   
351767                  200.0                0  INTERNET   
126377                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
104446                   8.865992   windows                   0   
269483                   4.654872     linux                   1   
4102                     3.720953     linux                   1   
351767                   2.912670   windows                   1   
126377                   2.283680     other                   1   

       device_fraud_count  month        x1        x2  
104446                  0      2 -0.245425  0.568811  
269483                  0      7  0.009336 -2.096682  
4102                    0      1  2.229616 -0.005823  
351767                  0      3 -0.193945  0.861207  
126377                  0      3 -2.440650  0.354986  

[5 rows x 29 columns]


 Últimas 5 filas:


id  fraud_bool  income  prev_address_months_count  \
8689    500233           0     0.3                         -1   
341572  764920           0     0.2                         -1   
93740   479316           0     0.8                         52   
34890   453943           0     0.8                         -1   
193209  607582           0     0.8                         -1   

        current_address_months_count  customer_age  days_since_request  \
8689                             190            30            0.024909   
341572                           308            20            0.042292   
93740                            156            40            0.014419   
34890                            384            50            0.018356   
193209                            43            50            0.032342   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
8689                 -0.848016           AB          1682  ...   
341572               -1.670041           AD          1669  ...   
93740                42.877023           AA           759  ...   
34890                43.489859           AA           894  ...   
193209               -1.276301           AB          1514  ...   

        proposed_credit_limit  foreign_request    source  \
8689                   1500.0                0  INTERNET   
341572                 1000.0                0  INTERNET   
93740                  1500.0                0  INTERNET   
34890                   200.0                0  INTERNET   
193209                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
8689                     4.043459     linux                   1   
341572                  12.412305     linux                   1   
93740                    1.407784   windows                   1   
34890                    6.288238     other                   0   
193209                   5.671012   windows                   0   

       device_fraud_count  month        x1        x2  
8689                    0      2 -0.972253 -0.164525  
341572                  0      2 -0.433835 -2.137383  
93740                   0      3 -1.108988  0.467937  
34890                   0      0 -0.870616 -0.221149  
193209                  0      1 -1.111151 -0.779366  

[5 rows x 29 columns]


 Muestreo aleatorio de 5 filas:


id  fraud_bool  income  prev_address_months_count  \
15920     4372           0     0.2                         -1   
120909   48357           0     0.9                         -1   
46005   540273           0     0.9                         -1   
258442  841268           0     0.4                         -1   
328227  611430           0     0.4                         -1   

        current_address_months_count  customer_age  days_since_request  \
15920                            105            60            0.025944   
120909                            76            50            0.006248   
46005                             33            40            0.008133   
258442                            28            50            1.885663   
328227                            25            40            0.024405   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
15920                -1.099298           AB           749  ...   
120909              101.655731           AA          2464  ...   
46005                -1.014150           AA           437  ...   
258442               -0.965720           AC           276  ...   
328227               -1.218249           AB           308  ...   

        proposed_credit_limit  foreign_request    source  \
15920                   200.0                0  INTERNET   
120909                 1500.0                0  INTERNET   
46005                   200.0                0  INTERNET   
258442                  200.0                0  INTERNET   
328227                 1500.0                0  INTERNET   

        session_length_in_minutes  device_os  keep_alive_session  \
15920                   32.999553    windows                   1   
120909                   4.751635  macintosh                   0   
46005                    1.457720      other                   0   
258442                   1.591574      linux                   1   
328227                  12.611629      other                   1   

       device_fraud_count  month        x1        x2  
15920                   0      7 -0.219916 -0.493999  
120909                  0      2  3.737853  1.295432  
46005                   0      7 -1.767114  1.194423  
258442                  0      2 -0.660358 -0.900845  
328227                  0      6  0.146351 -2.177451  

[5 rows x 29 columns]


 4.- Descripción numérica del Dataframe:



id     fraud_bool         income  prev_address_months_count  \
count  389782.000000  389782.000000  389782.000000              389782.000000   
mean   499848.005603       0.011081       0.578593                  14.725716   
std    288668.519647       0.104680       0.288225                  43.142125   
min         0.000000       0.000000       0.100000                  -1.000000   
25%    249505.500000       0.000000       0.300000                  -1.000000   
50%    499452.000000       0.000000       0.600000                  -1.000000   
75%    749926.750000       0.000000       0.800000                  -1.000000   
max    999998.000000       1.000000       0.900000                 382.000000   

       current_address_months_count   customer_age  days_since_request  \
count                 389782.000000  389782.000000        3.897820e+05   
mean                      99.266274      41.356066        8.982626e-01   
std                       94.051447      13.753951        4.971210e+00   
min                       -1.000000      10.000000        3.112791e-08   
25%                       27.000000      30.000000        7.473412e-03   
50%                       64.000000      50.000000        1.571523e-02   
75%                      154.000000      50.000000        2.699353e-02   
max                      429.000000      90.000000        7.635261e+01   

       intended_balcon_amount   zip_count_4w    velocity_6h  ...  \
count           389782.000000  389782.000000  389782.000000  ...   
mean                 8.571027    1518.623882    5492.044466  ...   
std                 20.560324     966.525083    2938.767195  ...   
min                -15.537329       1.000000    -174.109691  ...   
25%                 -1.179831     886.000000    3338.460577  ...   
50%                 -0.834662    1208.000000    5189.775212  ...   
75%                 -0.059201    1844.000000    7371.002553  ...   
max                112.702504    6650.000000   16754.959024  ...   

       bank_months_count  has_other_cards  proposed_credit_limit  \
count      389782.000000    389782.000000          389782.000000   
mean           11.139091         0.249768             550.722429   
std            12.131114         0.432879             506.326217   
min            -1.000000         0.000000             190.000000   
25%             1.000000         0.000000             200.000000   
50%             6.000000         0.000000             200.000000   
75%            25.000000         0.000000            1000.000000   
max            32.000000         1.000000            2100.000000   

       foreign_request  session_length_in_minutes  keep_alive_session  \
count    389782.000000              389782.000000       389782.000000   
mean          0.023898                   7.816942            0.555906   
std           0.152732                   8.242139            0.496865   
min           0.000000                  -1.000000            0.000000   
25%           0.000000                   3.149521            0.000000   
50%           0.000000                   5.248347            1.000000   
75%           0.000000                   9.358599            1.000000   
max           1.000000                  85.567848            1.000000   

       device_fraud_count          month             x1             x2  
count            389782.0  389782.000000  389782.000000  389782.000000  
mean                  0.0       3.655446       0.012611       0.010676  
std                   0.0       2.116485       1.012885       1.014503  
min                   0.0       0.000000      -4.602975      -4.846414  
25%                   0.0       2.000000      -0.669224      -0.672295  
50%                   0.0       4.000000       0.004778       0.003866  
75%                   0.0       5.000000       0.685736       0.684303  
max                   0.0       7.000000       5.518076       6.542492  

[8 rows x 24 columns]


 5.- Tipo de datos por columna:


id                                int64
fraud_bool                        int64
income                          float64
prev_address_months_count         int64
current_address_months_count      int64
customer_age                      int64
days_since_request              float64
intended_balcon_amount          float64
payment_type                     object
zip_count_4w                      int64
velocity_6h                     float64
velocity_24h                    float64
velocity_4w                     float64
bank_branch_count_8w              int64
employment_status                object
credit_risk_score                 int64
housing_status                   object
bank_months_count                 int64
has_other_cards                   int64
proposed_credit_limit           float64
foreign_request                   int64
source                           object
session_length_in_minutes       float64
device_os                        object
keep_alive_session                int64



 6.- Cantidad de valores nulos por columna:


id                              0
fraud_bool                      0
income                          0
prev_address_months_count       0
current_address_months_count    0
customer_age                    0
days_since_request              0
intended_balcon_amount          0
payment_type                    0
zip_count_4w                    0
velocity_6h                     0
velocity_24h                    0
velocity_4w                     0
bank_branch_count_8w            0
employment_status               0
credit_risk_score               0
housing_status                  0
bank_months_count               0
has_other_cards                 0
proposed_credit_limit           0
foreign_request                 0
source                          0
session_length_in_minutes       0
device_os                       0
keep_alive_session              0
device_fraud_count              0
month                           0
x1                              0
x2                              0
dtype: int64


 7.- Cantidad de valores nulos en total:


0


 8.- Cantidad de valores únicos por columna:


id                              389782
fraud_bool                           2
income                               9
prev_address_months_count          371
current_address_months_count       411
customer_age                         9
days_since_request              388087
intended_balcon_amount          389002
payment_type                         5
zip_count_4w                      6024
velocity_6h                     389583
velocity_24h                    389620
velocity_4w                     389479
bank_branch_count_8w              2280
employment_status                    7
credit_risk_score                  534
housing_status                       7
bank_months_count                   33
has_other_cards                      2
proposed_credit_limit               12
foreign_request                      2
source                               2
session_length_in_minutes       388566
device_os                            5
keep_alive_session                   2
device_fraud_count       

In [6]:
exploratory_data_analysis(df_2)

1.- El DataFrame tiene 747410 filas y 29 columnas

2.- El DataFrame esta compuesto por las siguientes columnas:  ['id', 'fraud_bool', 'income', 'prev_address_months_count', 'current_address_months_count', 'customer_age', 'days_since_request', 'intended_balcon_amount', 'payment_type', 'zip_count_4w', 'velocity_6h', 'velocity_24h', 'velocity_4w', 'bank_branch_count_8w', 'employment_status', 'credit_risk_score', 'housing_status', 'bank_months_count', 'has_other_cards', 'proposed_credit_limit', 'foreign_request', 'source', 'session_length_in_minutes', 'device_os', 'keep_alive_session', 'device_fraud_count', 'month', 'x1', 'x2'] 

3.- Ejemplos de filas del DataFrame:

Primeras 5 filas:


id  fraud_bool  income  prev_address_months_count  \
699954  303612           0     0.8                         -1   
701206  568066           0     0.9                         -1   
270304  893344           0     0.7                         -1   
346538  247688           0     0.8                         -1   
147297  674478           0     0.8                         -1   

        current_address_months_count  customer_age  days_since_request  \
699954                            21            40            0.010837   
701206                            70            50            0.711485   
270304                           133            50            0.006856   
346538                           207            50            0.026194   
147297                            24            50            0.012646   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
699954               -0.853976           AD           239  ...   
701206               25.137456           AA          1873  ...   
270304               -0.815330           AC           675  ...   
346538               -0.736615           AB           999  ...   
147297               10.744253           AA          3198  ...   

        proposed_credit_limit  foreign_request    source  \
699954                  200.0                0  INTERNET   
701206                  200.0                0  INTERNET   
270304                 1000.0                0  INTERNET   
346538                  200.0                0  INTERNET   
147297                  500.0                1  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
699954                  48.127390     other                   0   
701206                   8.792273   windows                   1   
270304                  15.916225     other                   1   
346538                   0.509942     linux                   1   
147297                  10.562945     other                   0   

       device_fraud_count  month        x1        x2  
699954                  0      3  0.236931  0.077061  
701206                  0      3  0.571536  1.151563  
270304                  0      7  0.224252 -0.590459  
346538                  0      6  0.648528 -0.315911  
147297                  0      6  0.041475  0.383332  

[5 rows x 29 columns]


 Últimas 5 filas:


id  fraud_bool  income  prev_address_months_count  \
626893   28360           0     0.8                         -1   
747411  428637           0     0.6                         -1   
121778   41856           0     0.7                         25   
568036  656236           0     0.6                         -1   
384611  988191           0     0.9                         -1   

        current_address_months_count  customer_age  days_since_request  \
626893                           137            30            0.018466   
747411                            60            50            0.026457   
121778                            18            30            1.426644   
568036                            85            60            0.008277   
384611                            56            30            0.006165   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
626893               49.088586           AA           915  ...   
747411               -0.949519           AD          3145  ...   
121778               10.441915           AA          1188  ...   
568036               -0.914454           AB          1021  ...   
384611               -1.103130           AD          1922  ...   

        proposed_credit_limit  foreign_request    source  \
626893                  500.0                0  INTERNET   
747411                  200.0                0  INTERNET   
121778                  200.0                0  INTERNET   
568036                  200.0                0  INTERNET   
384611                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
626893                   4.090987     other                   0   
747411                   9.759439   windows                   1   
121778                   2.899507     other                   1   
568036                   4.031749     linux                   0   
384611                   3.843607     linux                   0   

       device_fraud_count  month        x1        x2  
626893                  0      1 -0.563622 -0.204385  
747411                  0      5 -0.874186  0.310623  
121778                  0      3  1.642429 -0.026783  
568036                  0      5 -2.168143  0.745766  
384611                  0      4 -1.662247  1.576214  

[5 rows x 29 columns]


 Muestreo aleatorio de 5 filas:


id  fraud_bool  income  prev_address_months_count  \
605705  417593           0     0.5                         -1   
493632  184225           0     0.5                         -1   
56551   407351           0     0.4                         10   
484515  679507           0     0.1                         -1   
325487  258369           0     0.3                         79   

        current_address_months_count  customer_age  days_since_request  \
605705                           133            50            0.012431   
493632                           202            40            0.020040   
56551                             23            60            0.025470   
484515                           191            50            0.008552   
325487                           205            60            0.015450   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
605705               -0.416021           AD           914  ...   
493632               -0.815098           AB           691  ...   
56551                -1.161107           AB           968  ...   
484515               -1.271051           AC          1078  ...   
325487               -1.411951           AB           937  ...   

        proposed_credit_limit  foreign_request    source  \
605705                  200.0                0  INTERNET   
493632                  200.0                0  INTERNET   
56551                   200.0                0  INTERNET   
484515                  200.0                0  INTERNET   
325487                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
605705                   5.373880     other                   0   
493632                   3.670265     other                   1   
56551                   27.272793     linux                   0   
484515                  25.122497   windows                   0   
325487                   6.941133     linux                   0   

       device_fraud_count  month        x1        x2  
605705                  0      2  1.049818  1.088895  
493632                  0      1  0.494775 -1.131096  
56551                   0      3 -0.298899  0.495094  
484515                  0      2  1.671906 -1.061511  
325487                  0      1  0.198509 -0.831603  

[5 rows x 29 columns]


 4.- Descripción numérica del Dataframe:



id     fraud_bool         income  prev_address_months_count  \
count  747410.000000  747410.000000  747410.000000              747410.000000   
mean   500251.605708       0.010973       0.578524                  14.605949   
std    288750.620564       0.104174       0.288247                  42.896994   
min         3.000000       0.000000       0.100000                  -1.000000   
25%    250120.250000       0.000000       0.300000                  -1.000000   
50%    500827.500000       0.000000       0.600000                  -1.000000   
75%    750219.500000       0.000000       0.800000                  -1.000000   
max    999999.000000       1.000000       0.900000                 399.000000   

       current_address_months_count   customer_age  days_since_request  \
count                 747410.000000  747410.000000        7.474100e+05   
mean                      99.127609      41.337445        9.035222e-01   
std                       94.070215      13.772967        5.015799e+00   
min                       -1.000000      10.000000        3.112791e-08   
25%                       27.000000      30.000000        7.414064e-03   
50%                       64.000000      50.000000        1.562815e-02   
75%                      153.000000      50.000000        2.685132e-02   
max                      423.000000      90.000000        7.657750e+01   

       intended_balcon_amount   zip_count_4w    velocity_6h  ...  \
count           747410.000000  747410.000000  747410.000000  ...   
mean                 8.555753    1518.484254    5491.685202  ...   
std                 20.518142     965.950459    2940.707374  ...   
min                -14.981743       1.000000    -155.430730  ...   
25%                 -1.179983     886.000000    3336.506724  ...   
50%                 -0.833807    1209.000000    5193.782151  ...   
75%                 -0.063765    1845.000000    7366.452368  ...   
max                112.613538    6526.000000   16754.200917  ...   

       bank_months_count  has_other_cards  proposed_credit_limit  \
count      747410.000000    747410.000000          747410.000000   
mean           11.153443         0.249583             552.080317   
std            12.126832         0.432772             506.725127   
min            -1.000000         0.000000             190.000000   
25%             1.000000         0.000000             200.000000   
50%             6.000000         0.000000             200.000000   
75%            25.000000         0.000000            1000.000000   
max            32.000000         1.000000            2100.000000   

       foreign_request  session_length_in_minutes  keep_alive_session  \
count    747410.000000              747410.000000       747410.000000   
mean          0.023870                   7.817974            0.556420   
std           0.152646                   8.242699            0.496807   
min           0.000000                  -1.000000            0.000000   
25%           0.000000                   3.153643            0.000000   
50%           0.000000                   5.249156            1.000000   
75%           0.000000                   9.368660            1.000000   
max           1.000000                  85.161998            1.000000   

       device_fraud_count          month             x1             x2  
count            747410.0  747410.000000  747410.000000  747410.000000  
mean                  0.0       3.656616       0.015617       0.011188  
std                   0.0       2.117112       1.013224       1.012026  
min                   0.0       0.000000      -4.977864      -4.783370  
25%                   0.0       2.000000      -0.666256      -0.669171  
50%                   0.0       4.000000       0.010293       0.005562  
75%                   0.0       5.000000       0.687275       0.683390  
max                   0.0       7.000000       6.434867       6.284777  

[8 rows x 24 columns]


 5.- Tipo de datos por columna:


id                                int64
fraud_bool                        int64
income                          float64
prev_address_months_count         int64
current_address_months_count      int64
customer_age                      int64
days_since_request              float64
intended_balcon_amount          float64
payment_type                     object
zip_count_4w                      int64
velocity_6h                     float64
velocity_24h                    float64
velocity_4w                     float64
bank_branch_count_8w              int64
employment_status                object
credit_risk_score                 int64
housing_status                   object
bank_months_count                 int64
has_other_cards                   int64
proposed_credit_limit           float64
foreign_request                   int64
source                           object
session_length_in_minutes       float64
device_os                        object
keep_alive_session                int64



 6.- Cantidad de valores nulos por columna:


id                              0
fraud_bool                      0
income                          0
prev_address_months_count       0
current_address_months_count    0
customer_age                    0
days_since_request              0
intended_balcon_amount          0
payment_type                    0
zip_count_4w                    0
velocity_6h                     0
velocity_24h                    0
velocity_4w                     0
bank_branch_count_8w            0
employment_status               0
credit_risk_score               0
housing_status                  0
bank_months_count               0
has_other_cards                 0
proposed_credit_limit           0
foreign_request                 0
source                          0
session_length_in_minutes       0
device_os                       0
keep_alive_session              0
device_fraud_count              0
month                           0
x1                              0
x2                              0
dtype: int64


 7.- Cantidad de valores nulos en total:


0


 8.- Cantidad de valores únicos por columna:


id                              584118
fraud_bool                           2
income                               9
prev_address_months_count          370
current_address_months_count       413
customer_age                         9
days_since_request              580339
intended_balcon_amount          582360
payment_type                         5
zip_count_4w                      6139
velocity_6h                     583689
velocity_24h                    583786
velocity_4w                     583466
bank_branch_count_8w              2304
employment_status                    7
credit_risk_score                  540
housing_status                       7
bank_months_count                   32
has_other_cards                      2
proposed_credit_limit               12
foreign_request                      2
source                               2
session_length_in_minutes       581771
device_os                            5
keep_alive_session                   2
device_fraud_count       

In [7]:
exploratory_data_analysis(df_email_phone)

1.- El DataFrame tiene 934730 filas y 7 columnas

2.- El DataFrame esta compuesto por las siguientes columnas:  ['id', 'name_email_similarity', 'date_of_birth_distinct_emails_4w', 'email_is_free', 'device_distinct_emails_8w', 'phone_home_valid', 'phone_mobile_valid'] 

3.- Ejemplos de filas del DataFrame:

Primeras 5 filas:


id  name_email_similarity  date_of_birth_distinct_emails_4w  \
987231  624588               0.555653                                15   
79954   620810               0.849718                                 2   
567130  580633               0.110898                                 3   
500891    6379               0.671250                                15   
55399   366511               0.772932                                 7   

        email_is_free  device_distinct_emails_8w  phone_home_valid  \
987231              0                          1                 0   
79954               0                          1                 0   
567130              0                          1                 1   
500891              0                          1                 0   
55399               0                          1                 0   

        phone_mobile_valid  
987231                   1  
79954                    0  
567130                   0  
500891                   1  
55399                    1


 Últimas 5 filas:


id  name_email_similarity  date_of_birth_distinct_emails_4w  \
89156   437057               0.181835                                 3   
590568  492432               0.486547                                 4   
950861  109950               0.549603                                10   
656247  346972               0.063439                                 1   
986114  226280               0.204169                                10   

        email_is_free  device_distinct_emails_8w  phone_home_valid  \
89156               1                          1                 1   
590568              0                          1                 1   
950861              1                          1                 1   
656247              1                          1                 0   
986114              1                          1                 1   

        phone_mobile_valid  
89156                    0  
590568                   1  
950861                   1  
656247                   1  
986114                   1


 Muestreo aleatorio de 5 filas:


id  name_email_similarity  date_of_birth_distinct_emails_4w  \
520458  117599               0.693986                                 4   
707941  605255               0.716083                                12   
465172  273801               0.897040                                14   
517826  362328               0.648051                                 4   
996836   87468               0.882102                                 6   

        email_is_free  device_distinct_emails_8w  phone_home_valid  \
520458              0                          1                 0   
707941              0                          1                 0   
465172              1                          1                 0   
517826              1                          1                 1   
996836              1                          1                 0   

        phone_mobile_valid  
520458                   1  
707941                   1  
465172                   1  
517826                   1  
996836                   1


 4.- Descripción numérica del Dataframe:



id  name_email_similarity  date_of_birth_distinct_emails_4w  \
count  934730.000000           9.347300e+05                     934730.000000   
mean   500013.586976           4.876156e-01                          7.771928   
std    288635.775620           2.914544e-01                          4.815612   
min         0.000000           5.024707e-08                          0.000000   
25%    249989.500000           2.145474e-01                          4.000000   
50%    500075.500000           4.858929e-01                          7.000000   
75%    749965.750000           7.545305e-01                         11.000000   
max    999999.000000           1.000000e+00                         39.000000   

       email_is_free  device_distinct_emails_8w  phone_home_valid  \
count  934730.000000              934730.000000     934730.000000   
mean        0.518645                   1.022067          0.493399   
std         0.499653                   0.191637          0.499957   
min         0.000000                  -1.000000          0.000000   
25%         0.000000                   1.000000          0.000000   
50%         1.000000                   1.000000          0.000000   
75%         1.000000                   1.000000          1.000000   
max         1.000000                   2.000000          1.000000   

       phone_mobile_valid  
count       934730.000000  
mean             0.856989  
std              0.350085  
min              0.000000  
25%              1.000000  
50%              1.000000  
75%              1.000000  
max              1.000000


 5.- Tipo de datos por columna:


id                                    int64
name_email_similarity               float64
date_of_birth_distinct_emails_4w      int64
email_is_free                         int64
device_distinct_emails_8w             int64
phone_home_valid                      int64
phone_mobile_valid                    int64
dtype: object


 6.- Cantidad de valores nulos por columna:


id                                  0
name_email_similarity               0
date_of_birth_distinct_emails_4w    0
email_is_free                       0
device_distinct_emails_8w           0
phone_home_valid                    0
phone_mobile_valid                  0
dtype: int64


 7.- Cantidad de valores nulos en total:


0


 8.- Cantidad de valores únicos por columna:


id                                  934730
name_email_similarity               933577
date_of_birth_distinct_emails_4w        40
email_is_free                            2
device_distinct_emails_8w                4
phone_home_valid                         2
phone_mobile_valid                       2
dtype: int64

### 2.3 Unir dataframes

Ahora se procede a unir los dataframes con la función concat de pandas para poder unir sin limpiar los datos del dataframe 2 al dataframe 1.

In [8]:
merged_df = pd.concat([df_1, df_2])

In [9]:
exploratory_data_analysis(merged_df)

1.- El DataFrame tiene 1137192 filas y 29 columnas

2.- El DataFrame esta compuesto por las siguientes columnas:  ['id', 'fraud_bool', 'income', 'prev_address_months_count', 'current_address_months_count', 'customer_age', 'days_since_request', 'intended_balcon_amount', 'payment_type', 'zip_count_4w', 'velocity_6h', 'velocity_24h', 'velocity_4w', 'bank_branch_count_8w', 'employment_status', 'credit_risk_score', 'housing_status', 'bank_months_count', 'has_other_cards', 'proposed_credit_limit', 'foreign_request', 'source', 'session_length_in_minutes', 'device_os', 'keep_alive_session', 'device_fraud_count', 'month', 'x1', 'x2'] 

3.- Ejemplos de filas del DataFrame:

Primeras 5 filas:


id  fraud_bool  income  prev_address_months_count  \
104446  729517           0     0.7                         -1   
269483  149585           0     0.8                         -1   
4102     64486           0     0.9                         -1   
351767  825283           0     0.5                         -1   
126377    8308           0     0.9                         -1   

        current_address_months_count  customer_age  days_since_request  \
104446                           305            60            0.030059   
269483                           140            50            0.015659   
4102                             171            50            0.001409   
351767                            85            30            0.027292   
126377                            39            30            0.010945   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
104446               -1.599455           AC           990  ...   
269483                3.951994           AA          1269  ...   
4102                 28.159779           AB          4430  ...   
351767               -1.310498           AB          1698  ...   
126377               -1.450972           AC           569  ...   

        proposed_credit_limit  foreign_request    source  \
104446                  500.0                0  INTERNET   
269483                  200.0                0  INTERNET   
4102                   1500.0                0  INTERNET   
351767                  200.0                0  INTERNET   
126377                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
104446                   8.865992   windows                   0   
269483                   4.654872     linux                   1   
4102                     3.720953     linux                   1   
351767                   2.912670   windows                   1   
126377                   2.283680     other                   1   

       device_fraud_count  month        x1        x2  
104446                  0      2 -0.245425  0.568811  
269483                  0      7  0.009336 -2.096682  
4102                    0      1  2.229616 -0.005823  
351767                  0      3 -0.193945  0.861207  
126377                  0      3 -2.440650  0.354986  

[5 rows x 29 columns]


 Últimas 5 filas:


id  fraud_bool  income  prev_address_months_count  \
626893   28360           0     0.8                         -1   
747411  428637           0     0.6                         -1   
121778   41856           0     0.7                         25   
568036  656236           0     0.6                         -1   
384611  988191           0     0.9                         -1   

        current_address_months_count  customer_age  days_since_request  \
626893                           137            30            0.018466   
747411                            60            50            0.026457   
121778                            18            30            1.426644   
568036                            85            60            0.008277   
384611                            56            30            0.006165   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
626893               49.088586           AA           915  ...   
747411               -0.949519           AD          3145  ...   
121778               10.441915           AA          1188  ...   
568036               -0.914454           AB          1021  ...   
384611               -1.103130           AD          1922  ...   

        proposed_credit_limit  foreign_request    source  \
626893                  500.0                0  INTERNET   
747411                  200.0                0  INTERNET   
121778                  200.0                0  INTERNET   
568036                  200.0                0  INTERNET   
384611                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
626893                   4.090987     other                   0   
747411                   9.759439   windows                   1   
121778                   2.899507     other                   1   
568036                   4.031749     linux                   0   
384611                   3.843607     linux                   0   

       device_fraud_count  month        x1        x2  
626893                  0      1 -0.563622 -0.204385  
747411                  0      5 -0.874186  0.310623  
121778                  0      3  1.642429 -0.026783  
568036                  0      5 -2.168143  0.745766  
384611                  0      4 -1.662247  1.576214  

[5 rows x 29 columns]


 Muestreo aleatorio de 5 filas:


id  fraud_bool  income  prev_address_months_count  \
415335   96704           0     0.9                         32   
407049  111228           0     0.9                         -1   
776931  834296           0     0.7                         -1   
587076  653354           0     0.7                         -1   
1214    889396           0     0.7                        136   

        current_address_months_count  customer_age  days_since_request  \
415335                             6            30            0.001008   
407049                           154            60            0.021904   
776931                            32            50            0.018526   
587076                            51            50            0.014261   
1214                             141            70            0.023713   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
415335               -1.077306           AB           648  ...   
407049               -0.027753           AB          1427  ...   
776931               -0.517184           AB          1450  ...   
587076               -0.565723           AB          1064  ...   
1214                 99.520443           AA          1126  ...   

        proposed_credit_limit  foreign_request    source  \
415335                  200.0                0  INTERNET   
407049                 1500.0                0  INTERNET   
776931                  200.0                0  INTERNET   
587076                  200.0                0  INTERNET   
1214                    500.0                0  INTERNET   

        session_length_in_minutes  device_os  keep_alive_session  \
415335                   8.014311  macintosh                   0   
407049                   3.898471    windows                   0   
776931                   6.779432      linux                   1   
587076                  14.092194      linux                   1   
1214                     4.525431      other                   1   

       device_fraud_count  month        x1        x2  
415335                  0      5 -0.605840 -0.274778  
407049                  0      1  1.715679  1.108327  
776931                  0      3 -0.432506  0.425943  
587076                  0      4  0.097636  0.283442  
1214                    0      3 -0.390380  1.378258  

[5 rows x 29 columns]


 4.- Descripción numérica del Dataframe:



id    fraud_bool        income  prev_address_months_count  \
count  1.137192e+06  1.137192e+06  1.137192e+06               1.137192e+06   
mean   5.001133e+05  1.100957e-02  5.785479e-01               1.464700e+01   
std    2.887224e+05  1.043474e-01  2.882396e-01               4.298119e+01   
min    0.000000e+00  0.000000e+00  1.000000e-01              -1.000000e+00   
25%    2.499185e+05  0.000000e+00  3.000000e-01              -1.000000e+00   
50%    5.003475e+05  0.000000e+00  6.000000e-01              -1.000000e+00   
75%    7.501132e+05  0.000000e+00  8.000000e-01              -1.000000e+00   
max    9.999990e+05  1.000000e+00  9.000000e-01               3.990000e+02   

       current_address_months_count  customer_age  days_since_request  \
count                  1.137192e+06  1.137192e+06        1.137192e+06   
mean                   9.917514e+01  4.134383e+01        9.017195e-01   
std                    9.406376e+01  1.376645e+01        5.000559e+00   
min                   -1.000000e+00  1.000000e+01        3.112791e-08   
25%                    2.700000e+01  3.000000e+01        7.434452e-03   
50%                    6.400000e+01  5.000000e+01        1.565940e-02   
75%                    1.540000e+02  5.000000e+01        2.689542e-02   
max                    4.290000e+02  9.000000e+01        7.657750e+01   

       intended_balcon_amount  zip_count_4w   velocity_6h  ...  \
count            1.137192e+06  1.137192e+06  1.137192e+06  ...   
mean             8.560989e+00  1.518532e+03  5.491808e+03  ...   
std              2.053260e+01  9.661470e+02  2.940041e+03  ...   
min             -1.553733e+01  1.000000e+00 -1.741097e+02  ...   
25%             -1.179949e+00  8.860000e+02  3.337138e+03  ...   
50%             -8.341029e-01  1.209000e+03  5.192389e+03  ...   
75%             -6.162908e-02  1.845000e+03  7.368034e+03  ...   
max              1.127025e+02  6.650000e+03  1.675496e+04  ...   

       bank_months_count  has_other_cards  proposed_credit_limit  \
count       1.137192e+06     1.137192e+06           1.137192e+06   
mean        1.114852e+01     2.496465e-01           5.516149e+02   
std         1.212830e+01     4.328086e-01           5.065886e+02   
min        -1.000000e+00     0.000000e+00           1.900000e+02   
25%         1.000000e+00     0.000000e+00           2.000000e+02   
50%         6.000000e+00     0.000000e+00           2.000000e+02   
75%         2.500000e+01     0.000000e+00           1.000000e+03   
max         3.200000e+01     1.000000e+00           2.100000e+03   

       foreign_request  session_length_in_minutes  keep_alive_session  \
count     1.137192e+06               1.137192e+06        1.137192e+06   
mean      2.387987e-02               7.817620e+00        5.562438e-01   
std       1.526750e-01               8.242504e+00        4.968268e-01   
min       0.000000e+00              -1.000000e+00        0.000000e+00   
25%       0.000000e+00               3.152311e+00        0.000000e+00   
50%       0.000000e+00               5.248896e+00        1.000000e+00   
75%       0.000000e+00               9.365151e+00        1.000000e+00   
max       1.000000e+00               8.556785e+01        1.000000e+00   

       device_fraud_count         month            x1            x2  
count           1137192.0  1.137192e+06  1.137192e+06  1.137192e+06  
mean                  0.0  3.656215e+00  1.458640e-02  1.101237e-02  
std                   0.0  2.116896e+00  1.013109e+00  1.012876e+00  
min                   0.0  0.000000e+00 -4.977864e+00 -4.846414e+00  
25%                   0.0  2.000000e+00 -6.672525e-01 -6.702569e-01  
50%                   0.0  4.000000e+00  8.456920e-03  4.949004e-03  
75%                   0.0  5.000000e+00  6.867563e-01  6.837339e-01  
max                   0.0  7.000000e+00  6.434867e+00  6.542492e+00  

[8 rows x 24 columns]


 5.- Tipo de datos por columna:


id                                int64
fraud_bool                        int64
income                          float64
prev_address_months_count         int64
current_address_months_count      int64
customer_age                      int64
days_since_request              float64
intended_balcon_amount          float64
payment_type                     object
zip_count_4w                      int64
velocity_6h                     float64
velocity_24h                    float64
velocity_4w                     float64
bank_branch_count_8w              int64
employment_status                object
credit_risk_score                 int64
housing_status                   object
bank_months_count                 int64
has_other_cards                   int64
proposed_credit_limit           float64
foreign_request                   int64
source                           object
session_length_in_minutes       float64
device_os                        object
keep_alive_session                int64



 6.- Cantidad de valores nulos por columna:


id                              0
fraud_bool                      0
income                          0
prev_address_months_count       0
current_address_months_count    0
customer_age                    0
days_since_request              0
intended_balcon_amount          0
payment_type                    0
zip_count_4w                    0
velocity_6h                     0
velocity_24h                    0
velocity_4w                     0
bank_branch_count_8w            0
employment_status               0
credit_risk_score               0
housing_status                  0
bank_months_count               0
has_other_cards                 0
proposed_credit_limit           0
foreign_request                 0
source                          0
session_length_in_minutes       0
device_os                       0
keep_alive_session              0
device_fraud_count              0
month                           0
x1                              0
x2                              0
dtype: int64


 7.- Cantidad de valores nulos en total:


0


 8.- Cantidad de valores únicos por columna:


id                              934858
fraud_bool                           2
income                               9
prev_address_months_count          373
current_address_months_count       416
customer_age                         9
days_since_request              925125
intended_balcon_amount          930349
payment_type                         5
zip_count_4w                      6238
velocity_6h                     933695
velocity_24h                    933964
velocity_4w                     933273
bank_branch_count_8w              2320
employment_status                    7
credit_risk_score                  544
housing_status                       7
bank_months_count                   33
has_other_cards                      2
proposed_credit_limit               12
foreign_request                      2
source                               2
session_length_in_minutes       930312
device_os                            5
keep_alive_session                   2
device_fraud_count       

### 2.4 Unir nuevas variables

Luego, para unir los dos dataframes, es decir el dataframe que contiene a df_1 y df_2, con el dataframe que contiene números de teléfono y emails, lo que se utiliza es la estrategia **outer join** para poder utilizar la función merge de pandas, ya que 

In [10]:
final_df = pd.merge(merged_df, df_email_phone, how="outer", on='id')

### 2.5 Verificar nulos y duplicados, generar explicaciones sobre sus fuentes y proponer soluciones.



Para poder analizar el dataframe final, vamos a hacer un análisis exploratorio primero para poder ver de qué se compone este.

In [11]:
exploratory_data_analysis(final_df)

1.- El DataFrame tiene 1198151 filas y 35 columnas

2.- El DataFrame esta compuesto por las siguientes columnas:  ['id', 'fraud_bool', 'income', 'prev_address_months_count', 'current_address_months_count', 'customer_age', 'days_since_request', 'intended_balcon_amount', 'payment_type', 'zip_count_4w', 'velocity_6h', 'velocity_24h', 'velocity_4w', 'bank_branch_count_8w', 'employment_status', 'credit_risk_score', 'housing_status', 'bank_months_count', 'has_other_cards', 'proposed_credit_limit', 'foreign_request', 'source', 'session_length_in_minutes', 'device_os', 'keep_alive_session', 'device_fraud_count', 'month', 'x1', 'x2', 'name_email_similarity', 'date_of_birth_distinct_emails_4w', 'email_is_free', 'device_distinct_emails_8w', 'phone_home_valid', 'phone_mobile_valid'] 

3.- Ejemplos de filas del DataFrame:

Primeras 5 filas:


id  fraud_bool  income  prev_address_months_count  \
0  729517         0.0     0.7                       -1.0   
1  729517         0.0     0.7                       -1.0   
2  149585         0.0     0.8                       -1.0   
3  149585         0.0     0.8                       -1.0   
4   64486         0.0     0.9                       -1.0   

   current_address_months_count  customer_age  days_since_request  \
0                         305.0          60.0            0.030059   
1                         305.0          60.0            0.030059   
2                         140.0          50.0            0.015659   
3                         140.0          50.0            0.015659   
4                         171.0          50.0            0.001409   

   intended_balcon_amount payment_type  zip_count_4w  ...  device_fraud_count  \
0               -1.599455           AC         990.0  ...                 0.0   
1               -1.599455           AC         990.0  ...                 0.0   
2                3.951994           AA        1269.0  ...                 0.0   
3                3.951994           AA        1269.0  ...                 0.0   
4               28.159779           AB        4430.0  ...                 0.0   

   month        x1        x2 name_email_similarity  \
0    2.0 -0.245425  0.568811              0.883485   
1    2.0 -0.245425  0.568811              0.883485   
2    7.0  0.009336 -2.096682              0.113208   
3    7.0  0.009336 -2.096682              0.113208   
4    1.0  2.229616 -0.005823              0.792797   

   date_of_birth_distinct_emails_4w email_is_free  device_distinct_emails_8w  \
0                               2.0           1.0                        1.0   
1                               2.0           1.0                        1.0   
2                               4.0           1.0                        1.0   
3                               4.0           1.0                        1.0   
4                               4.0           0.0                        1.0   

   phone_home_valid  phone_mobile_valid  
0               0.0                 1.0  
1               0.0                 1.0  
2               1.0                 1.0  
3               1.0                 1.0  
4               0.0                 1.0  

[5 rows x 35 columns]


 Últimas 5 filas:


id  fraud_bool  income  prev_address_months_count  \
1198146   36622         NaN     NaN                        NaN   
1198147  507501         NaN     NaN                        NaN   
1198148  468237         NaN     NaN                        NaN   
1198149  469817         NaN     NaN                        NaN   
1198150  549209         NaN     NaN                        NaN   

         current_address_months_count  customer_age  days_since_request  \
1198146                           NaN           NaN                 NaN   
1198147                           NaN           NaN                 NaN   
1198148                           NaN           NaN                 NaN   
1198149                           NaN           NaN                 NaN   
1198150                           NaN           NaN                 NaN   

         intended_balcon_amount payment_type  zip_count_4w  ...  \
1198146                     NaN          NaN           NaN  ...   
1198147                     NaN          NaN           NaN  ...   
1198148                     NaN          NaN           NaN  ...   
1198149                     NaN          NaN           NaN  ...   
1198150                     NaN          NaN           NaN  ...   

         device_fraud_count  month  x1  x2 name_email_similarity  \
1198146                 NaN    NaN NaN NaN              0.488584   
1198147                 NaN    NaN NaN NaN              0.615749   
1198148                 NaN    NaN NaN NaN              0.493888   
1198149                 NaN    NaN NaN NaN              0.006752   
1198150                 NaN    NaN NaN NaN              0.761055   

         date_of_birth_distinct_emails_4w email_is_free  \
1198146                              26.0           1.0   
1198147                               3.0           0.0   
1198148                               3.0           1.0   
1198149                               2.0           0.0   
1198150                               7.0           0.0   

         device_distinct_emails_8w  phone_home_valid  phone_mobile_valid  
1198146                        2.0               0.0                 1.0  
1198147                        1.0               1.0                 1.0  
1198148                        1.0               1.0                 0.0  
1198149                        1.0               1.0                 0.0  
1198150                        1.0               1.0                 1.0  

[5 rows x 35 columns]


 Muestreo aleatorio de 5 filas:


id  fraud_bool  income  prev_address_months_count  \
1025370  921563         0.0     0.8                       31.0   
577264   656837         0.0     0.3                       -1.0   
418754   899204         0.0     0.1                       -1.0   
798960   126875         0.0     0.6                       -1.0   
310809   614891         0.0     0.7                       -1.0   

         current_address_months_count  customer_age  days_since_request  \
1025370                          41.0          50.0            0.010293   
577264                           47.0          20.0            2.725315   
418754                          238.0          50.0            0.010398   
798960                          178.0          50.0            0.006577   
310809                           19.0          30.0            0.034323   

         intended_balcon_amount payment_type  zip_count_4w  ...  \
1025370               -0.747391           AC        2949.0  ...   
577264                -0.655903           AC         885.0  ...   
418754                -1.046026           AC         952.0  ...   
798960                -1.100230           AD        2739.0  ...   
310809                -0.893249           AD        1054.0  ...   

         device_fraud_count  month        x1        x2 name_email_similarity  \
1025370                 0.0    2.0 -1.445507  0.345125              0.834096   
577264                  0.0    2.0 -0.590834  0.115526              0.091846   
418754                  0.0    1.0  0.950216  0.726334              0.478994   
798960                  0.0    0.0 -0.500562 -1.019951              0.781791   
310809                  0.0    4.0  0.230611 -0.685465              0.491760   

         date_of_birth_distinct_emails_4w email_is_free  \
1025370                               8.0           1.0   
577264                                9.0           0.0   
418754                                3.0           0.0   
798960                                7.0           0.0   
310809                               10.0           0.0   

         device_distinct_emails_8w  phone_home_valid  phone_mobile_valid  
1025370                        1.0               0.0                 0.0  
577264                         1.0               0.0                 1.0  
418754                         2.0               0.0                 1.0  
798960                         1.0               0.0                 1.0  
310809                         1.0               1.0                 1.0  

[5 rows x 35 columns]


 4.- Descripción numérica del Dataframe:



id    fraud_bool        income  prev_address_months_count  \
count  1.198151e+06  1.137192e+06  1.137192e+06               1.137192e+06   
mean   5.001283e+05  1.100957e-02  5.785479e-01               1.464700e+01   
std    2.887415e+05  1.043474e-01  2.882396e-01               4.298119e+01   
min    0.000000e+00  0.000000e+00  1.000000e-01              -1.000000e+00   
25%    2.499515e+05  0.000000e+00  3.000000e-01              -1.000000e+00   
50%    5.002730e+05  0.000000e+00  6.000000e-01              -1.000000e+00   
75%    7.502525e+05  0.000000e+00  8.000000e-01              -1.000000e+00   
max    9.999990e+05  1.000000e+00  9.000000e-01               3.990000e+02   

       current_address_months_count  customer_age  days_since_request  \
count                  1.137192e+06  1.137192e+06        1.137192e+06   
mean                   9.917514e+01  4.134383e+01        9.017195e-01   
std                    9.406376e+01  1.376645e+01        5.000559e+00   
min                   -1.000000e+00  1.000000e+01        3.112791e-08   
25%                    2.700000e+01  3.000000e+01        7.434452e-03   
50%                    6.400000e+01  5.000000e+01        1.565940e-02   
75%                    1.540000e+02  5.000000e+01        2.689542e-02   
max                    4.290000e+02  9.000000e+01        7.657750e+01   

       intended_balcon_amount  zip_count_4w   velocity_6h  ...  \
count            1.137192e+06  1.137192e+06  1.137192e+06  ...   
mean             8.560989e+00  1.518532e+03  5.491808e+03  ...   
std              2.053260e+01  9.661470e+02  2.940041e+03  ...   
min             -1.553733e+01  1.000000e+00 -1.741097e+02  ...   
25%             -1.179949e+00  8.860000e+02  3.337138e+03  ...   
50%             -8.341029e-01  1.209000e+03  5.192389e+03  ...   
75%             -6.162908e-02  1.845000e+03  7.368034e+03  ...   
max              1.127025e+02  6.650000e+03  1.675496e+04  ...   

       device_fraud_count         month            x1            x2  \
count           1137192.0  1.137192e+06  1.137192e+06  1.137192e+06   
mean                  0.0  3.656215e+00  1.458640e-02  1.101237e-02   
std                   0.0  2.116896e+00  1.013109e+00  1.012876e+00   
min                   0.0  0.000000e+00 -4.977864e+00 -4.846414e+00   
25%                   0.0  2.000000e+00 -6.672525e-01 -6.702569e-01   
50%                   0.0  4.000000e+00  8.456920e-03  4.949004e-03   
75%                   0.0  5.000000e+00  6.867563e-01  6.837339e-01   
max                   0.0  7.000000e+00  6.434867e+00  6.542492e+00   

       name_email_similarity  date_of_birth_distinct_emails_4w  email_is_free  \
count           1.118424e+06                      1.118424e+06   1.118424e+06   
mean            4.875620e-01                      7.776952e+00   5.181559e-01   
std             2.914812e-01                      4.814635e+00   4.996705e-01   
min             5.024707e-08                      0.000000e+00   0.000000e+00   
25%             2.144731e-01                      4.000000e+00   0.000000e+00   
50%             4.857949e-01                      7.000000e+00   1.000000e+00   
75%             7.544987e-01                      1.100000e+01   1.000000e+00   
max             1.000000e+00                      3.900000e+01   1.000000e+00   

       device_distinct_emails_8w  phone_home_valid  phone_mobile_valid  
count               1.118424e+06      1.118424e+06        1.118424e+06  
mean                1.022179e+00      4.933549e-01        8.569478e-01  
std                 1.915275e-01      4.999561e-01        3.501262e-01  
min                -1.000000e+00      0.000000e+00        0.000000e+00  
25%                 1.000000e+00      0.000000e+00        1.000000e+00  
50%                 1.000000e+00      0.000000e+00        1.000000e+00  
75%                 1.000000e+00      1.000000e+00        1.000000e+00  
max                 2.000000e+00      1.000000e+00        1.000000e+00  

[8 rows x 30 columns]


 5.- Tipo de datos por columna:


id                                    int64
fraud_bool                          float64
income                              float64
prev_address_months_count           float64
current_address_months_count        float64
customer_age                        float64
days_since_request                  float64
intended_balcon_amount              float64
payment_type                         object
zip_count_4w                        float64
velocity_6h                         float64
velocity_24h                        float64
velocity_4w                         float64
bank_branch_count_8w                float64
employment_status                    object
credit_risk_score                   float64
housing_status                       object
bank_months_count                   float64
has_other_cards                     float64
proposed_credit_limit               float64
foreign_request                     float64
source                               object
session_length_in_minutes       


 6.- Cantidad de valores nulos por columna:


id                                      0
fraud_bool                          60959
income                              60959
prev_address_months_count           60959
current_address_months_count        60959
customer_age                        60959
days_since_request                  60959
intended_balcon_amount              60959
payment_type                        60959
zip_count_4w                        60959
velocity_6h                         60959
velocity_24h                        60959
velocity_4w                         60959
bank_branch_count_8w                60959
employment_status                   60959
credit_risk_score                   60959
housing_status                      60959
bank_months_count                   60959
has_other_cards                     60959
proposed_credit_limit               60959
foreign_request                     60959
source                              60959
session_length_in_minutes           60959
device_os                         


 7.- Cantidad de valores nulos en total:


2185214


 8.- Cantidad de valores únicos por columna:


id                                  995817
fraud_bool                               2
income                                   9
prev_address_months_count              373
current_address_months_count           416
customer_age                             9
days_since_request                  925125
intended_balcon_amount              930349
payment_type                             5
zip_count_4w                          6238
velocity_6h                         933695
velocity_24h                        933964
velocity_4w                         933273
bank_branch_count_8w                  2320
employment_status                        7
credit_risk_score                      544
housing_status                           7
bank_months_count                       33
has_other_cards                          2
proposed_credit_limit                   12
foreign_request                          2
source                                   2
session_length_in_minutes           930312
device_os  

En este caso se puede ver que el dataframe está compuesto de 1.198.151 filas, sin embargo, posee 995.817 *id* únicas, por lo que claramente se pueden ver que hay *ids* de clientes repetidas que van a tener que ser tratadas. También se pueden ver una cantidad de NaN en el dataframe (sin contar id), con 60.959 valores nulos para todas las columnas de los dos primeros dataframes, y 79.727 para las columnas del dataframe de teléfonos y emails. Revisando numéricamente el dataframe, hay valores que llaman la atención para los clientes de este banco, y que se plantea el siguiente tratamiento:
* **prev_address_months_count**: Se tiene que esta variable va desde -1 a 399, cuando en la descripción de este dice que va desde -1 a 380, por lo que se va a tener que acotar el rango. Cabe destacar que esos -1 representan valores nulos.
* **device_fraud_count**: Esta variable se encuentra repleta de 0, por lo que no aporta información y será descartada.

En cuanto a los valores nulos, como representan como máximo el 6,65% de los datos, se van a descartar para poder tener limpio el dataframe.

Se ven casos particulares en columnas para saber si existen valores que no deberian estar. 

In [12]:

print('Se esperan 5 valores (sin contar na)', final_df['payment_type'].unique())
print('Se esperan 7 valores (sin contar na)', final_df['employment_status'].unique())
print('Se esperan 2 valores (sin contar na)', final_df['email_is_free'].unique())
print('Se esperan 7 valores (sin contar na)', final_df['housing_status'].unique())
print('Se esperan 2 valores (sin contar na)', final_df['phone_home_valid'].unique())
print('Se esperan 2 valores (sin contar na)', final_df['phone_mobile_valid'].unique())
print('Se esperan 2 valores (sin contar na)', final_df['has_other_cards'].unique())
print('Se esperan 2 valores (sin contar na)', final_df['foreign_request'].unique())
print('Se esperan 2 valores (sin contar na)', final_df['source'].unique())
print('Se esperan 5 valores (sin contar na)', final_df['device_os'].unique())
print('Se esperan 2 valores (sin contar na)', final_df['keep_alive_session'].unique())
print('Se esperan 2 valores (sin contar na)', final_df['fraud_bool'].unique())


Se esperan 5 valores (sin contar na) ['AC' 'AA' 'AB' 'AD' 'AE' nan]
Se esperan 7 valores (sin contar na) ['CC' 'CF' 'CA' 'CB' 'CE' 'CD' 'CG' nan]
Se esperan 2 valores (sin contar na) [ 1.  0. nan]
Se esperan 7 valores (sin contar na) ['BB' 'BA' 'BE' 'BC' 'BD' 'BF' 'BG' nan]
Se esperan 2 valores (sin contar na) [ 0.  1. nan]
Se esperan 2 valores (sin contar na) [ 1.  0. nan]
Se esperan 2 valores (sin contar na) [ 0.  1. nan]
Se esperan 2 valores (sin contar na) [ 0.  1. nan]
Se esperan 2 valores (sin contar na) ['INTERNET' 'TELEAPP' nan]
Se esperan 5 valores (sin contar na) ['windows' 'linux' 'other' 'macintosh' 'x11' nan]
Se esperan 2 valores (sin contar na) [ 0.  1. nan]
Se esperan 2 valores (sin contar na) [ 0.  1. nan]


Se puede ver que en las varaibles cateogoricas o binarias no existen valores que se escapan de lo esperado excepto por lo NA.

Por ultimo se ve la cantidad de Na por columna en aquellas columnas que sus Na estan representadas por un numero negativo.

In [13]:
columnas_de_interes = ['prev_address_months_count', 
                       'current_address_months_count', 
                       'intended_balcon_amount', 
                       'bank_months_count', 
                       'session_length_in_minutes',
                       'device_distinct_emails_8w']

for columna in columnas_de_interes:
    num_negative_values = final_df[columna].lt(0).sum() # lt(0) devuelve True para valores negativos
    total_values = len(final_df[columna])
    percentage_negative = (num_negative_values / total_values) * 100
    print(f"Porcentaje de valores negativos en {columna}: {round(percentage_negative,1)}%")

Porcentaje de valores negativos en prev_address_months_count: 72.3%
Porcentaje de valores negativos en current_address_months_count: 0.3%
Porcentaje de valores negativos en intended_balcon_amount: 71.3%
Porcentaje de valores negativos en bank_months_count: 23.4%
Porcentaje de valores negativos en session_length_in_minutes: 0.2%
Porcentaje de valores negativos en device_distinct_emails_8w: 0.0%


Como existe una gran cantidad de valores nulos en prev_address_months_count y intended_balcon_amount, se borraran estas columnas ya que no aportan mucha informacion. 

### 2.6 Limpiar


Para elegir el ID que se conservará del conjunto de datos, se dará prioridad al ID del primer dataframe, que se considera más confiable que el segundo. Para esto se hacen borran los duplicados y se mantiene el primer registro encontrado que es el del primer dataset. 

In [14]:
filas_iniciales = len(final_df)

final_df.dropna(inplace=True)
conteo_drop_na = len(final_df)

final_df.drop_duplicates(subset=['id'], inplace=True, keep= 'first')
conteo_duplicados = len(final_df)

print("Filas removidas con NA: ", filas_iniciales - conteo_drop_na)
print("Filas removidas con duplicados: ", conteo_drop_na - conteo_duplicados)

final_df.drop('device_fraud_count', axis=1, inplace= True)
final_df.drop('intended_balcon_amount', axis=1, inplace= True)
final_df.drop('prev_address_months_count', axis=1, inplace= True)

Filas removidas con NA:  140686
Filas removidas con duplicados:  183694


Se ajusta cada columna por el rango de los valores a los que pertenece.

In [15]:
filas_iniciales = len(final_df)

final_df.query('income >= 0.1 and income <= 0.9', inplace=True)
final_df.query('name_email_similarity >= 0 and name_email_similarity <= 1', inplace=True)
final_df.query('current_address_months_count >= 0 and current_address_months_count <= 429', inplace=True)
final_df.query('customer_age >= 10 and customer_age <= 90', inplace=True)
final_df.query('days_since_request >= 0 and days_since_request <= 79', inplace=True)
final_df.query('zip_count_4w >= 1 and zip_count_4w <= 6830', inplace=True)
final_df.query('velocity_6h >= -175 and velocity_6h <= 16818', inplace=True)
final_df.query('velocity_24h >= 1297 and velocity_24h <= 9586', inplace=True)
final_df.query('velocity_4w >= 2825 and velocity_4w <= 7020', inplace=True)
final_df.query('bank_branch_count_8w >= 0 and bank_branch_count_8w <= 2404', inplace=True)
final_df.query('date_of_birth_distinct_emails_4w >= 0 and date_of_birth_distinct_emails_4w <= 39', inplace=True)
final_df.query('credit_risk_score >= -191 and credit_risk_score <= 389', inplace=True)
final_df.query('bank_months_count >= 0 and bank_months_count <= 32', inplace=True)
final_df.query('proposed_credit_limit >= 200 and proposed_credit_limit <= 2000', inplace=True)
final_df.query('session_length_in_minutes >= -1 and session_length_in_minutes <= 107', inplace=True)
final_df.query('device_distinct_emails_8w >= 0 and device_distinct_emails_8w <= 2', inplace=True)
final_df.query('month >= 0 and month <= 7', inplace=True)

filas_queried = len(final_df)

print("Filas removidas con query: ", filas_iniciales - filas_queried)

Filas removidas con query:  218462


Al buscar valores dentro de los rangos esperados, se perdieron 218462 registros, lo cual es una gran cantidad de estos. Sin embargo aun queda una gran cantidad en la base de datos.

### 2.7 Segmentar usuarios

Para un mejor análisis en el trabajo siguiente, se van a segmentar a los usuarios en 4 categorías:
* Joven (Menor a 18 años)
* Adulto-Joven (18 - 26 años)
* Adulto (27- 59 años)
* Persona Mayor (60 años o más)

In [16]:
conditions = [
    (final_df['customer_age'] < 18),
    (final_df['customer_age'] >= 18) & (final_df['customer_age'] <= 26),
    (final_df['customer_age'] >= 27) & (final_df['customer_age'] <= 59),
    (final_df['customer_age'] >= 60)
    ]

values = ['Joven', 'Adulto-Joven', 'Adulto', 'Persona Mayor']

final_df['segmentacion_etaria'] = np.select(conditions, values)

In [17]:
final_df.head()

id  fraud_bool  income  current_address_months_count  customer_age  \
2   149585         0.0     0.8                         140.0          50.0   
4    64486         0.0     0.9                         171.0          50.0   
5   825283         0.0     0.5                          85.0          30.0   
9   891636         0.0     0.6                         137.0          60.0   
11  395034         0.0     0.4                         156.0          40.0   

    days_since_request payment_type  zip_count_4w  velocity_6h  velocity_24h  \
2             0.015659           AA        1269.0  2844.260015   2789.827355   
4             0.001409           AB        4430.0  5854.065790   4210.039095   
5             0.027292           AB        1698.0  7938.525709   4963.194619   
9             0.046158           AA        1512.0  7669.494145   4251.257682   
11            0.032288           AA         941.0  6073.856953   3653.085358   

    ...  month        x1        x2  name_email_similarity  \
2   ...    7.0  0.009336 -2.096682               0.113208   
4   ...    1.0  2.229616 -0.005823               0.792797   
5   ...    3.0 -0.193945  0.861207               0.865082   
9   ...    2.0  2.192960  0.484068               0.383238   
11  ...    6.0 -0.647299 -0.319544               0.903835   

   date_of_birth_distinct_emails_4w  email_is_free  device_distinct_emails_8w  \
2                               4.0            1.0                        1.0   
4                               4.0            0.0                        1.0   
5                              10.0            0.0                        1.0   
9                               2.0            1.0                        1.0   
11                              5.0            1.0                        1.0   

    phone_home_valid  phone_mobile_valid segmentacion_etaria  
2                1.0                 1.0              Adulto  
4                0.0                 1.0              Adulto  
5                0.0                 1.0              Adulto  
9                1.0                 1.0       Persona Mayor  
11               1.0                 1.0              Adulto  

[5 rows x 33 columns]

## 3. Análisis Exploratorio de Datos

In [18]:
from ydata_profiling import ProfileReport

In [19]:
profile = ProfileReport(final_df, title="EDA")
profile.to_file("eda.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

### 3.1 Análisis del EDA

El resultado del analisis hecho por la libreria ydata_profiling da los siguientes resultados:

> Respecto a los valores NA, ya no se encuentra ninguno de estos en el dataset. Y los ids, se tiene un unico id por fila. 

Ahora se hara un analisis de cada variable del eda adjunto en el html:

> fraud_bool: Se ve que esta variable binaria tiene un desbalance gigante, donde los datos de 

## 4. Preprocesamiento

### 4.1 Declarar `ColumnTransformer`

In [20]:
numeric_transformer = Pipeline(steps=[
    ('scaler', StandardScaler())])

# Columnas categoricas
categorical_transformer = OneHotEncoder(handle_unknown='ignore')

# Columnas binarias
binary_transformer = OrdinalEncoder()

# Tipo de cada columna
numeric_features = ['income', 'current_address_months_count', 'customer_age',
                    'days_since_request', 'zip_count_4w', 'velocity_6h', 'velocity_24h',
                    'velocity_4w', 'bank_branch_count_8w', 'credit_risk_score', 'bank_months_count', 'proposed_credit_limit',
                    'session_length_in_minutes', 'month', 'x1', 'x2', 'name_email_similarity',
                    'date_of_birth_distinct_emails_4w', 'device_distinct_emails_8w']

categorical_features = ['payment_type', 'employment_status', 'housing_status', 'source', 'device_os', 'segmentacion_etaria']

binary_features = ['has_other_cards', 'foreign_request', 'keep_alive_session', 'email_is_free', 'phone_home_valid',
                   'phone_mobile_valid']

# Se define el column transformer
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features),
        ('bin', binary_transformer, binary_features)
    ],
    remainder='passthrough' # Para 'id' y 'fraud_bool' que no se deben transformar
)



### 4.2 Transformar datos

In [21]:
# Aplica la transformacion
data = preprocessor.fit_transform(final_df)

# Obtiene el nombre de las columnas
name_columns = preprocessor.get_feature_names_out()

# Lo guarda en un dataframe de pandas
data = pd.DataFrame(data, columns = name_columns)

## 5. Visualización en Baja Dimensionalidad

### 5.1 Muestrear dataframe

In [75]:
muestra = data.sample(n=25000, random_state=11)
features = muestra.drop(columns=['remainder__fraud_bool', 'remainder__id'])

### 5.2 Proyectar y agregar proyecciones al dataframe de muestreo

In [76]:
pca = PCA(n_components=2)
pca = pca.fit_transform(features)
pca_df = pd.DataFrame(data = pca, columns = ['principal component 1', 'principal component 2'])

pca_df = pca_df.reset_index(drop=True)
muestra = muestra.reset_index(drop=True)
pca_df['fraud_bool'] = muestra[['remainder__fraud_bool']].astype(int).astype('category')
pca_df['Mayor'] = muestra['cat__segmentacion_etaria_Persona Mayor'].astype('int')
pca_df['Adulto'] = muestra['cat__segmentacion_etaria_Adulto'].astype('int')
pca_df['Adulto-Joven'] = muestra['cat__segmentacion_etaria_Adulto-Joven'].astype('int')
pca_df['Joven'] = muestra['cat__segmentacion_etaria_Joven'].astype('int')



### 5.3 Visualizar según rangos etarios

In [77]:
# Adulto Mayor
fig_pca_adulto_mayor = px.scatter(pca_df[pca_df['Mayor'] == 1], 
                     x='principal component 1', 
                     y='principal component 2', 
                     color='fraud_bool', 
                     title='PCA Plot Adulto Mayor', 
                     color_discrete_sequence=['blue', 'red'])

fig_pca_adulto_mayor.show()

# Adulto
fig_pca_adulto = px.scatter(pca_df[pca_df['Adulto'] == 1], 
                     x='principal component 1', 
                     y='principal component 2', 
                     color='fraud_bool', 
                     title='PCA Plot Adulto', 
                     color_discrete_sequence=['blue', 'red'])

fig_pca_adulto.show()

# Adulto-Joven
fig_pca_adulto_joven = px.scatter(pca_df[pca_df['Adulto-Joven'] == 1], 
                     x='principal component 1', 
                     y='principal component 2', 
                     color='fraud_bool', 
                     title='PCA Plot Adulto-Joven', 
                     color_discrete_sequence=['blue', 'red'])

fig_pca_adulto_joven.show()

# Joven
fig_pca_joven = px.scatter(pca_df[pca_df['Joven'] == 1], 
                     x='principal component 1', 
                     y='principal component 2', 
                     color='fraud_bool', 
                     title='PCA Plot Joven', 
                     color_discrete_sequence=['blue', 'red'])

fig_pca_joven.show()


## 6. Anomalías

### 6.1 Implementar detector de anomalías sobre dataframe de muestreo

In [26]:
...

Ellipsis

### 6.2 Agregar resultados a dataframe de muestreo

In [27]:
...

Ellipsis

### 6.3 Visualizar según rangos etarios

In [28]:
...

Ellipsis

### 6.4 Calcular ratios y responder

In [29]:
...

Ellipsis